pkg/operator/finalizer.go

```diff
@@ -0,0 +1,174 @@
+	if err != nil {
+		return fmt.Errorf("unable to get previous config state: %s", err)
+	}
+
+	err = generate.RemoveConfigState(configState)
+	if err != nil {
+		return fmt.Errorf("unable to remove previous config state: %s", err)
+	}
+
+	return nil
+}
+
+func (h *Handler) finalizeResources(o *regopapi.ImageRegistry) error {
+	defer utilruntime.HandleCrash()
+
+	if o.ObjectMeta.DeletionTimestamp == nil {
+		return nil
+	}
+
+	finalizers := []string{}
+	for _, v := range o.ObjectMeta.Finalizers {
+		if v != parameters.ImageRegistryOperatorResourceFinalizer {
+			finalizers = append(finalizers, v)
+		}
+	}
+
+	if len(finalizers) == len(o.ObjectMeta.Finalizers) {
+		return nil
+	}
+
+	logrus.Infof("finalizing %s %s/%s", o.GetObjectKind().GroupVersionKind().Kind, o.Namespace, o.Name)
+
+	err := h.RemoveResources(o)
+	if err != nil {
+		return err
+	}
+
+	cr := o
+	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		if cr == nil {
+			cr = &regopapi.ImageRegistry{
+				TypeMeta: metav1.TypeMeta{
+					APIVersion: o.TypeMeta.APIVersion,
+					Kind:       o.TypeMeta.Kind,
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      o.ObjectMeta.Name,
+					Namespace: o.ObjectMeta.Namespace,
+				},
+			}
+
+			err := sdk.Get(cr)
+			if err != nil {
+				return fmt.Errorf("failed to get custom resource %s/%s: %s", o.ObjectMeta.Namespace, o.ObjectMeta.Name, err)
+			}
+
+			finalizers = []string{}
+			for _, v := range cr.ObjectMeta.Finalizers {
+				if v != parameters.ImageRegistryOperatorResourceFinalizer {
+					finalizers = append(finalizers, v)
+				}
+			}
+		}
+
+		cr.ObjectMeta.Finalizers = finalizers
+
+		err := sdk.Update(cr)
+		if err != nil {
+			cr = nil
+			return err
+		}
+		return nil
+	})
+
+	if err != nil {
+		return fmt.Errorf("unable to update finalizers in custom resource %s/%s: %s", o.ObjectMeta.Namespace, o.ObjectMeta.Name, err)
+	}
+
+	// These errors may indicate a transient error that we can retry in tests.
+	errorFuncs := []func(error) bool{
+		kerrors.IsInternalError,
+		kerrors.IsTimeout,
+		kerrors.IsServerTimeout,
+		kerrors.IsTooManyRequests,
+		utilnet.IsProbableEOF,
+		utilnet.IsConnectionReset,
+	}
+
+	retryTime := 3 * time.Second
+
+	err = wait.PollInfinite(retryTime, func() (stop bool, err error) {
+		cr = &regopapi.ImageRegistry{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: o.TypeMeta.APIVersion,
+				Kind:       o.TypeMeta.Kind,
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      o.ObjectMeta.Name,
+				Namespace: o.ObjectMeta.Namespace,
+			},
+		}
+
+		err = sdk.Get(cr)
+		if err == nil {
+			return
+		}
+
+		if !kerrors.IsNotFound(err) {
+			for _, isRetryError := range errorFuncs {
+				if isRetryError(err) {
+					return
+				}
+			}
+
+			// If the error sends the Retry-After header, we respect it as an explicit confirmation we should retry.
+			if delaySeconds, shouldRetry := kerrors.SuggestsClientDelay(err); shouldRetry {
+				delayTime := time.Duration(delaySeconds)
```

Comment:
time.Duration represents the time in nanoseconds. You should multiply the result by time.Second.

Reply:
Ooops. Cut-and-Paste error. Sorry. Fixed.